src/libcore/iter.rs

```diff
@@ -8,51 +8,278 @@
+//! used in idiomatic Rust code, so it's worth becoming familiar with them.
 //!
-//! This module defines Rust's core iteration trait. The `Iterator` trait has
-//! one unimplemented method, `next`. All other methods are derived through
-//! default methods to perform operations such as `zip`, `chain`, `enumerate`,
-//! and `fold`.
+//! Before explaining more, let's talk about how this module is structured:
 //!
-//! The goal of this module is to unify iteration across all containers in Rust.
-//! An iterator can be considered as a state machine which is used to track
-//! which element will be yielded next.
+//! # Organization
 //!
-//! There are various extensions also defined in this module to assist with
-//! various types of iteration, such as the `DoubleEndedIterator` for iterating
-//! in reverse, the `FromIterator` trait for creating a container from an
-//! iterator, and much more.
+//! This module is largely organized by type:
 //!
-//! # Rust's `for` loop
+//! * [Traits] are the core portion: these traits define what kind of iterators exist and what you
+//!   can do with them. The methods of these traits are worth putting some extra study time into.
+//! * [Functions] provide some helpful ways to create some basic iterators.
+//! * [Structs] are often the return types of the various methods on this module's traits. You'll
+//!   usually want to look at the method that creates the `struct`, rather than the `struct`
+//!   itself. For more detail about why, see '[Implementing Iterator](#implementing-iterator)'.
 //!
-//! The special syntax used by rust's `for` loop is based around the
-//! `IntoIterator` trait defined in this module. `for` loops can be viewed as a
-//! syntactical expansion into a `loop`, for example, the `for` loop in this
-//! example is essentially translated to the `loop` below.
+//! [Traits]: http://doc.rust-lang.org/std/iter/index.html#traits
+//! [Functions]: http://doc.rust-lang.org/std/iter/index.html#functions
+//! [Structs]: http://doc.rust-lang.org/std/iter/index.html#structs
 //!
+//! That's it! Let's dig into iterators.
+//!
+//! # Iterator
+//!
+//! The heart and soul of this module is the [`Iterator`] trait. The core of [`Iterator`] looks
+//! like this:
+//!
+//! ```
+//! trait Iterator {
+//!     type Item;
+//!     fn next(&mut self) -> Option<Self::Item>;
+//! }
+//! ```
+//!
+//! An iterator has a method, [`next()`], which when called, returns an [`Option`]`<Item>`.
+//! [`next()`] will return `Some(Item)` as long as there are elements, and once they've all
+//! been exhausted, will return `None` to indicate that iteration is finished. Individual iterators
+//! may choose to resume iteration, and so calling [`next()`] again may or may not eventually start
+//! returning `Some(Item)` again at some point.
+//!
+//! [`Iterator`]'s full definition includes a number of other methods as well, but they are all
+//! built on top of [`next()`], and so you get the other methods for free.
```

Comment:
Perhaps the word "default methods" could be mentioned here? (just in the sense of it's a standard Rust concept)